File: services/ftp-service/internal/idgen/idgen.go
// Package idgen 为 ftp-service 提供会话 ID 等应用层唯一标识生成。
package idgen

import (
	"errors"
	"strconv"
	"sync"
	"time"
)

const (
	workerBits       = 10
	sequenceBits     = 12
	maxWorkerID      = (1 << workerBits) - 1
	maxSequence      = (1 << sequenceBits) - 1
	timestampShift   = workerBits + sequenceBits
	workerShift      = sequenceBits
	epochMillisecond int64 = 1704067200000
)

var (
	ErrWorkerIDOutOfRange = errors.New("idgen: worker id out of range")
	ErrClockBackwards     = errors.New("idgen: clock moved backwards")
)

// Generator 签发全局唯一 ID。
type Generator interface {
	NextID() (int64, error)
	NextString() (string, error)
}

// Snowflake 实现 Generator 接口。
type Snowflake struct {
	mu        sync.Mutex
	workerID  int64
	lastMilli int64
	sequence  int64
}

// NewSnowflake 创建 Snowflake 生成器。
func NewSnowflake(workerID int64) (*Snowflake, error) {
	if workerID < 0 || workerID > maxWorkerID {
		return nil, ErrWorkerIDOutOfRange
	}
	return &Snowflake{workerID: workerID}, nil
}

func (s *Snowflake) NextID() (int64, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	now := time.Now().UnixMilli()
	if now < s.lastMilli {
		return 0, ErrClockBackwards
	}
	if now == s.lastMilli {
		s.sequence = (s.sequence + 1) & maxSequence
		if s.sequence == 0 {
			for now <= s.lastMilli {
				now = time.Now().UnixMilli()
			}
		}
	} else {
		s.sequence = 0
	}
	s.lastMilli = now

	id := ((now - epochMillisecond) << timestampShift) |
		(s.workerID << workerShift) |
		s.sequence
	return id, nil
}

func (s *Snowflake) NextString() (string, error) {
	id, err := s.NextID()
	if err != nil {
		return "", err
	}
	return strconv.FormatInt(id, 10), nil
}

File: services/sftp-service/src/idgen.rs
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use thiserror::Error;

const WORKER_BITS: u64 = 10;
const SEQUENCE_BITS: u64 = 12;
const MAX_WORKER_ID: i64 = (1 << WORKER_BITS) - 1;
const MAX_SEQUENCE: i64 = (1 << SEQUENCE_BITS) - 1;
const TIMESTAMP_SHIFT: u64 = WORKER_BITS + SEQUENCE_BITS;
const WORKER_SHIFT: u64 = SEQUENCE_BITS;
const EPOCH_MILLISECOND: i64 = 1_704_067_200_000; // 2024-01-01 UTC

#[derive(Debug, Error)]
pub enum IdGenError {
    #[error("idgen: worker id out of range")]
    WorkerIdOutOfRange,
    #[error("idgen: clock moved backwards")]
    ClockBackwards,
}

/// Snowflake 签发全局唯一 ID（与 Go ftp-service / ssh-service idgen 位布局一致）。
pub struct Snowflake {
    worker_id: i64,
    last_milli: AtomicI64,
    sequence: Mutex<i64>,
}

impl Snowflake {
    pub fn new(worker_id: i64) -> Result<Self, IdGenError> {
        if !(0..=MAX_WORKER_ID).contains(&worker_id) {
            return Err(IdGenError::WorkerIdOutOfRange);
        }
        Ok(Self {
            worker_id,
            last_milli: AtomicI64::new(0),
            sequence: Mutex::new(0),
        })
    }

    pub fn next_id(&self) -> Result<i64, IdGenError> {
        let now = now_millis();
        let mut last = self.last_milli.load(Ordering::Acquire);
        if now < last {
            return Err(IdGenError::ClockBackwards);
        }

        let mut seq_guard = self.sequence.lock().unwrap();
        let mut sequence = *seq_guard;
        if now == last {
            sequence = (sequence + 1) & MAX_SEQUENCE;
            if sequence == 0 {
                let mut spin = now;
                while spin <= last {
                    spin = now_millis();
                }
                last = spin;
                self.last_milli.store(last, Ordering::Release);
            }
        } else {
            sequence = 0;
            last = now;
            self.last_milli.store(last, Ordering::Release);
        }
        *seq_guard = sequence;

        let id = ((last - EPOCH_MILLISECOND) << TIMESTAMP_SHIFT)
            | (self.worker_id << WORKER_SHIFT)
            | sequence;
        Ok(id)
    }

    pub fn next_string(&self) -> Result<String, IdGenError> {
        Ok(self.next_id()?.to_string())
    }
}

fn now_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("clock")
        .as_millis() as i64
}
